Message Lock renewal

lawrencegripper opened the pull request:
Here is an initial port of the ion code I had to renew locks over to the library, I doubt this will run in it's current form hence the WIP tag. 

I'm currently blocked as I need to take the newer version of vcabbage with my fixes for the marshalling code and that needs a change to the common lib:

```
dep ensure -add pack.ag/amqp@v0.7.2
Fetching sources...

Solving failure: No versions of github.com/Azure/azure-amqp-common-go met constraints:
	v0.6.0: Could not introduce github.com/Azure/azure-amqp-common-go@v0.6.0, as it has a dependency on pack.ag/amqp with constraint ^0.6.0, which has no overlap with existing constraint ^0.7.2 from (root)
	v0.7.0: Could not introduce github.com/Azure/azure-amqp-common-go@v0.7.0, as it is not allowed by constraint ^0.6.0 from project github.com/Azure/azure-service-bus-go.

```

Issue: #34 

Todo:

- [x] Unit test
- [x] Integration test
- [x] Tidy up

Great to get any early thoughts and recommendations

devigned commented:
after #36 gets merged, you should be all set

devigned commented on lockrenewal.go:
```diff
@@ -0,0 +1,113 @@
+	e.renewMessageLockMutex.Lock()
+	defer e.renewMessageLockMutex.Unlock()
+
+	sender, receiver, err := e.createManagementChannels(ctx, e.Name)
```
Check out the RPC stuff here: https://github.com/Azure/azure-amqp-common-go/blob/master/rpc/rpc.go.

The RPC functionality is used for Event Hubs management requests as seen here: https://github.com/Azure/azure-event-hubs-go/blob/master/amqp_mgmt.go.

I think this would reduce your code a bit.

devigned commented:
I agree with all of the points in your proposal and how you have structured the progression of the work.

I don't particularly like the consumer needing to call RenewLock as I think this will be fraught with pain. I would much prefer to give the consumer the ability to manage lock renewal themselves via the Entity, but make it the 1% path. I think 99% of customers will use auto-renewal and let the library manage renewing locks. It's a detail that a consumer should not need to think about.

devigned commented:
@lawrencegripper thank you for your hard work. I was afraid this was going to be a bit of a journey.

As for v0.7.4 of AMQP, you should just need to run `dep ensure -update` to use the latest version of vcabbage/amqp (under 1.0).

devigned commented:
Nah, you should be good to go. The constraint on the version in common allows for updates in downstream projects. I just ran it on my local machine and then did a `git diff` on the Gopkg.lock file. You can see the results below. The updated version dependency is updated from v0.7.3 to v0.7.4. 

```
 [[projects]]
-  digest = "1:efd7f4a2c3339485190142d3b83052207aaadf6e4e724fdbac79a5de0a8120c5"
+  digest = "1:4739800ec9c08baa4abefa388a31ff74d9e3b3a9cf112fe312332c6e6c3d119d"
   name = "pack.ag/amqp"
   packages = [
     ".",
     "internal/testconn",
   ]
   pruneopts = "UT"
-  revision = "f04efab52fd05cbd1b25280ae54423fc7b12e267"
-  version = "v0.7.3"
+  revision = "271d6282f59cd10184fef147de589cce8ad69c93"
+  version = "v0.7.4"

```

Let me know if you run into issues.

lawrencegripper commented:
Sooo, I got a bit distracted when I got to running the integration tests. I *love* that there is a Terraform template to create the infrastructure and it's using '.env' files to pull in the test environment variables. 

I've added some glue into the Terraform template and the Makefile to hook these two together. Terraform will create all the test infrastructure (including a SP scoped to the test resource group) then output these. Makefile picks up from here and takes these outputs and writes them into the '.env' file in the repo root. 

The end result is running `make test` (assuming you have TF install and are logged into AZ CLI) will deploy the test infrastructure and then wire up all the environment vars so the test can run without any input from the user. 

I hope this is useful, although not strictly related to this PR - Happy to pull out separately if that would be better.

lawrencegripper commented:
@devigned so I moved over the RPC stuff which is awesome - one issue is that the [`status-code`](https://github.com/lawrencegripper/azure-service-bus-go/blob/lg/lockrenewal/vendor/github.com/Azure/azure-amqp-common-go/rpc/rpc.go#L43) fields used between eventhub and SB aren't consistent in their RPC messages. 

SB gives me a [`statusCode`](https://github.com/lawrencegripper/ion/blob/lg/quickstart-doc/internal/pkg/servicebus/amqpconnection.go#L258) property :(

Easiest way to handle it, I can see, would be to try one and if that's not present try the next. Something like:

``` golang
	statusCode, ok := res.ApplicationProperties[statusCodeKey].(int32)
	if !ok {
		statusCode, ok = res.ApplicationProperties["statusCode"].(int32)
		if !ok {
			return nil, errors.New("status codes was not found on rpc message")
		}
	}
```

I'm going to try this out now as a quick fix, let me know your thoughts.

devigned commented:
@lawrencegripper check out the work here from @marstr https://github.com/Azure/azure-amqp-common-go/pull/25

I think that should fix up the issue you both have run into.

devigned commented:
Per the terraform stuff, I'll defer to @marstr. He was the person behind that work. 

Based on conversations he and I have had about improving the contribution and on-boarding experience for the project, I think your suggestions about how to further integrate TF would be welcomed enthusiastically.

lawrencegripper commented:
Ah nice - thanks @marstr https://github.com/Azure/azure-amqp-common-go/pull/25 looks like exactly what I need. I'll finish up my tests with the temp hack and then can update to that once it's merged. 

I'm running a bit light on time to tackle this any further due to other time pressures - been a bit of a journey. 

I'm hoping to get my tests working proving renewal on a queue behaves correctly then fix up all the review feedback but at that point I'm probably going to need to hand off the PR (sadly).

lawrencegripper commented:
Hopefully good for a review now, I've gone a little snow blind to code style trying to get this fixed up today so any and all pointers much appreciated. 

```
=== RUN   TestSB/TestQueueSendReceiveWithLock/SimpleSendReceiveWithLock
unhandled error link 15d7ff8e-e934-44de-afdf-de51d200a42f: status code 410 and description: tempfix
--- PASS: TestSB (295.47s)
    --- PASS: TestSB/TestQueueSendReceiveWithLock (289.70s)
        --- PASS: TestSB/TestQueueSendReceiveWithLock/SimpleSendReceiveWithLock (289.70s)
        	lockrenewal_test.go:53: Sending/receiving 27 messages
PASS
```

^ with manual fix to RPC status codes for now.

marstr commented on namespace.go:
```diff
@@ -133,3 +133,7 @@ func (ns *Namespace) getHTTPSHostURI() string {
 	return ns.getAMQPHostURI() + entityPath
 }
+
+func (ns *Namespace) getEntityManagementPath(entityPath string) string {
```
I [added getters](https://github.com/marstr/azure-service-bus-go/blob/1bd35f2be21e2a0bcd306513fc12663bde81d92b/queue.go#L292) to the `entity` type for this in my branch.

I can spin up a new PR to add those in quick so that you can take a dependency on them instead of adding this method.

marstr replied:
Closing the loop, I opened #60 for my side of this.

lawrencegripper replied:
Looking at the PR your getter returns the full path (including the amqp host URI). The code I've written is just after the entityName + "/$managment" (Sorry my bad on naming I started out using the full path before I realized it was incorrect then didn't fix up the method naming. 

What would be the best way to approach this?

lawrencegripper replied:
Done

marstr commented on lockrenewal.go:
```diff
@@ -0,0 +1,77 @@
+	}
+
+	e.renewMessageLockMutex.Lock()
+	defer e.renewMessageLockMutex.Unlock()
```
One benefit of moving this method to the message granularity is that we should be able to add a mutex at the message level instead of having a single instance of a lock renewer running at a time. Is there anything that you know of that challenges that assumption?

marstr commented on lockrenewal.go:
```diff
@@ -0,0 +1,77 @@
+)
+
+const (
+	serviceBuslockRenewalOperationName = "com.microsoft:renew-lock"
```
I'm not sure exactly how I'd like to organize these constants, but I too have been finding a lot of need for this. Maybe we could spin this and other const definitions into an `operations.go` file that just collects operation names and other related constants?

lawrencegripper replied:
Sounds like a good plan I'll create the file and relocate these.

marstr commented on lockrenewal.go:
```diff
@@ -0,0 +1,77 @@
+)
+
+//RenewLocks renews the locks on messages provided
+func (e *entity) RenewLocks(ctx context.Context, messages []*Message) error {
```
Looking at the PR history, I see that @devigned gave what amounts to contradicting feedback. I guess my $0.02 on this are:

- Between automatic and manual message lock renewal, I'm not sure which way ought to be the 99% case. Automatic will protect people who have taken too long to process for legitimate reasons, like message processing just being computationally expensive. However, it potentially hides bugs that have caused deadlock. At the end of the day, I don't have a strong opinion here, we just need to document our decision.
- I believe the `receiver` should be the one doing the automatic message lock renewal. Passing in a `receiverOption` that configures it to do so. It then parlays that into calling `Message.RenewLock` periodically as appropriate.

lawrencegripper replied:
Yeah it's a tough one to know either way. I'd usually look to take prior art and consistency between the SB SDK's as a guide - in the existing .NET SDK I *believe* that autorenewal is the default behavior and I've never seen people handling this manually. 

Because of this I'd structured the `RenewLock` method to sit on the entity and accepted batches of messages to allow `Receiver` to use this call efficiently when the AutoLocking functionality is added. By efficiently I'm thinking each receiver keep a track of all active messages and renew the batch with a single RPC round trip vs having 100 open messages and having to make 100 RPC calls - I can see a common usecase having a large number of messages active at one time. This also feeds into the locking comment too, as renewal occurs in batch infrequently a mutex per entity seems like a fair tradeoff. 

Then the method is also available on the entity for those who want to use it manually with autorenewal turned off. 

What do you think?

marstr replied:
You've convinced me. :)

I think down the road, I may add a method off of Message so that you can either RenewLock directly off the message, or continue to use the batch method that you're adding here.

lawrencegripper commented on .vscode/launch.json:
```diff
@@ -0,0 +1,35 @@
+{
```
Happy to remove, the reason I added was as it can be super useful for people coming to the repo to pick up easily and F5 in VSCode. As the file isn't .go it will be stripped out by dep when the SDK is imported so felt like a reasonable addition but can see the other side that it's only useful to those using VSCode so happy to remove. 

(We used quite a bit to capture linting settings, common tasks and debugging steps in one place - for example: https://github.com/lawrencegripper/ion/tree/master/.vscode )

lawrencegripper replied:
Done

lawrencegripper commented on azuredeploy.tf:
```diff
@@ -3,22 +3,35 @@ variable "location" {
 }
 
-variable "servicebus_name" {
+variable "servicebus_name_prefix" {
```
Am I ok to leave the commit in this PR as well and have Git work it out when merging or would you like me to do some interactive rebasing to remove it from this PR?

lawrencegripper commented on lockrenewal.go:
```diff
@@ -0,0 +1,77 @@
+	"context"
+	"fmt"
+	"github.com/Azure/azure-amqp-common-go/log"
+	"github.com/Azure/azure-amqp-common-go/rpc"
```
Done

lawrencegripper commented:
@marstr Hopefully all set now, I've moved to v1.1.1 of the common lib with your status code fix and merged in the master branch to pick up the changes from #60.

I think the only remaining point is around API design question, I've added my thoughts to your comment - let me know how you'd like to proceed.

lawrencegripper commented:
Re-ran integration test after these changes and all looks good except one example fails but it looks like this had been failing for some time and is unrelated to the PR. 

```
=== RUN   ExampleMessage_ScheduleAt
--- FAIL: ExampleMessage_ScheduleAt (9.28s)
got:
Received outside the expected window.
want:
Received when expected!
```

Integration tests
```
11:33 $ go test -v -timeout 20m .
=== RUN   TestSB
=== RUN   TestSB/TestAMQPMessageToMessage
=== RUN   TestSB/TestCreateNamespaceFromConnectionString
=== RUN   TestSB/TestEntryUnmarshal
=== RUN   TestSB/TestFeedUnmarshal
=== RUN   TestSB/TestMapStructureEncode
=== RUN   TestSB/TestMessageToAMQPMessage
=== RUN   TestSB/TestQueueClient
=== RUN   TestSB/TestQueueClient/MessageProperties
=== RUN   TestSB/TestQueueClient/Retry
=== RUN   TestSB/TestQueueClient/SimpleSend
=== RUN   TestSB/TestQueueClient/DuplicateDetection
=== RUN   TestSB/TestQueueEntryUnmarshal
=== RUN   TestSB/TestQueueManagement
=== RUN   TestSB/TestQueueManagement/TestQueueWithPartitioning
=== RUN   TestSB/TestQueueManagement/TestQueueDefaultSettings
=== RUN   TestSB/TestQueueManagement/TestQueueWithDeadLetteringOnMessageExpiration
=== RUN   TestSB/TestQueueManagement/TestQueueWithMaxSizeInMegabytes
=== RUN   TestSB/TestQueueManagement/TestQueueWithDuplicateDetection
=== RUN   TestSB/TestQueueManagement/TestQueueWithMessageTimeToLive
=== RUN   TestSB/TestQueueManagement/TestQueueWithLockDuration
=== RUN   TestSB/TestQueueManagement/TestQueueWithAutoDeleteOnIdle
=== RUN   TestSB/TestQueueManagement/TestQueueWithRequiredSessions
=== RUN   TestSB/TestQueueManagementPopulatedQueue
=== RUN   TestSB/TestQueueManagementPopulatedQueue/TestCountDetails
=== RUN   TestSB/TestQueueManagementReads
=== RUN   TestSB/TestQueueManagementReads/TestGetQueue
=== RUN   TestSB/TestQueueManagementReads/TestListQueues
=== RUN   TestSB/TestQueueManagementWrites
=== RUN   TestSB/TestQueueManagementWrites/TestPutDefaultQueue
=== RUN   TestSB/TestQueueSendReceiveWithLock
=== RUN   TestSB/TestQueueSendReceiveWithLock/SimpleSendReceiveWithLock
=== RUN   TestSB/TestQueueUnmarshal
=== RUN   TestSB/TestQueueWithReceiveAndDelete
=== RUN   TestSB/TestQueueWithReceiveAndDelete/SimpleSendAndReceive
=== RUN   TestSB/TestSubscriptionClient
=== RUN   TestSB/TestSubscriptionClient/SimpleReceive
context canceled
=== RUN   TestSB/TestSubscriptionClient/ReceiveOne
=== RUN   TestSB/TestSubscriptionEntryUnmarshal
=== RUN   TestSB/TestSubscriptionManagement
=== RUN   TestSB/TestSubscriptionManagement/TestSubscriptionWithAutoDeleteOnIdle
=== RUN   TestSB/TestSubscriptionManagement/TestSubscriptionWithRequiredSessions
=== RUN   TestSB/TestSubscriptionManagement/TestSubscriptionWithDeadLetteringOnMessageExpiration
=== RUN   TestSB/TestSubscriptionManagement/TestSubscriptionWithMessageTimeToLive
=== RUN   TestSB/TestSubscriptionManagement/TestSubscriptionWithLockDuration
=== RUN   TestSB/TestSubscriptionManagement/TestSubscriptionWithBatchedOperations
=== RUN   TestSB/TestSubscriptionManagement/TestSubscriptionDefaultSettings
=== RUN   TestSB/TestSubscriptionManagementWrites
=== RUN   TestSB/TestSubscriptionManagementWrites/TestPutDefaultSubscription
=== RUN   TestSB/TestSubscriptionUnmarshal
=== RUN   TestSB/TestTopic
=== RUN   TestSB/TestTopic/SimpleSend
=== RUN   TestSB/TestTopicEntryUnmarshal
=== RUN   TestSB/TestTopicManagement
=== RUN   TestSB/TestTopicManagement/TopicWithExpress
=== RUN   TestSB/TestTopicManagement/TopicWithDuplicateDetection
=== RUN   TestSB/TestTopicManagement/TopicWithAutoDeleteOnIdle
=== RUN   TestSB/TestTopicManagement/TopicWithTimeToLive
=== RUN   TestSB/TestTopicManagement/TopicWithBatchOperations
=== RUN   TestSB/TestTopicManagement/TopicWithMaxSizeInMegabytes
=== RUN   TestSB/TestTopicManagement/DefaultTopicCreation
=== RUN   TestSB/TestTopicManagement/TopicWithPartitioning
=== RUN   TestSB/TestTopicManagement/TopicWithOrdering
=== RUN   TestSB/TestTopicManagementReads
=== RUN   TestSB/TestTopicManagementReads/TestGetTopic
=== RUN   TestSB/TestTopicManagementReads/TestListTopics
=== RUN   TestSB/TestTopicManagementWrites
=== RUN   TestSB/TestTopicManagementWrites/TestPutDefaultTopic
=== RUN   TestSB/TestTopicUnmarshal
--- PASS: TestSB (493.76s)
    --- PASS: TestSB/TestAMQPMessageToMessage (0.00s)
    --- PASS: TestSB/TestCreateNamespaceFromConnectionString (0.00s)
    --- PASS: TestSB/TestEntryUnmarshal (0.00s)
    --- PASS: TestSB/TestFeedUnmarshal (0.00s)
    --- PASS: TestSB/TestMapStructureEncode (0.00s)
    --- PASS: TestSB/TestMessageToAMQPMessage (0.00s)
    --- PASS: TestSB/TestQueueClient (53.75s)
        --- PASS: TestSB/TestQueueClient/MessageProperties (15.83s)
        --- PASS: TestSB/TestQueueClient/Retry (15.21s)
        	queue_test.go:732: try 1 out of 10, message count was 1, not 0
        --- PASS: TestSB/TestQueueClient/SimpleSend (7.90s)
        --- PASS: TestSB/TestQueueClient/DuplicateDetection (14.81s)
        	queue_test.go:732: try 1 out of 10, message count was 1, not 0
    --- PASS: TestSB/TestQueueEntryUnmarshal (0.00s)
    --- PASS: TestSB/TestQueueManagement (20.80s)
        --- PASS: TestSB/TestQueueManagement/TestQueueWithPartitioning (2.56s)
        --- PASS: TestSB/TestQueueManagement/TestQueueDefaultSettings (2.24s)
        --- PASS: TestSB/TestQueueManagement/TestQueueWithDeadLetteringOnMessageExpiration (2.24s)
        --- PASS: TestSB/TestQueueManagement/TestQueueWithMaxSizeInMegabytes (2.23s)
        --- PASS: TestSB/TestQueueManagement/TestQueueWithDuplicateDetection (2.15s)
        --- PASS: TestSB/TestQueueManagement/TestQueueWithMessageTimeToLive (2.33s)
        --- PASS: TestSB/TestQueueManagement/TestQueueWithLockDuration (2.24s)
        --- PASS: TestSB/TestQueueManagement/TestQueueWithAutoDeleteOnIdle (2.56s)
        --- PASS: TestSB/TestQueueManagement/TestQueueWithRequiredSessions (2.24s)
    --- PASS: TestSB/TestQueueManagementPopulatedQueue (8.38s)
        --- PASS: TestSB/TestQueueManagementPopulatedQueue/TestCountDetails (8.38s)
    --- PASS: TestSB/TestQueueManagementReads (5.32s)
        --- PASS: TestSB/TestQueueManagementReads/TestGetQueue (0.22s)
        --- PASS: TestSB/TestQueueManagementReads/TestListQueues (0.68s)
    --- PASS: TestSB/TestQueueManagementWrites (1.90s)
        --- PASS: TestSB/TestQueueManagementWrites/TestPutDefaultQueue (1.90s)
    --- PASS: TestSB/TestQueueSendReceiveWithLock (263.50s)
        --- PASS: TestSB/TestQueueSendReceiveWithLock/SimpleSendReceiveWithLock (263.50s)
        	lockrenewal_test.go:53: Sending/receiving 57 messages
    --- PASS: TestSB/TestQueueUnmarshal (0.00s)
    --- PASS: TestSB/TestQueueWithReceiveAndDelete (15.80s)
        --- PASS: TestSB/TestQueueWithReceiveAndDelete/SimpleSendAndReceive (15.80s)
        	queue_test.go:587: Sending/receiving 31 messages
    --- PASS: TestSB/TestSubscriptionClient (33.46s)
        --- PASS: TestSB/TestSubscriptionClient/SimpleReceive (17.19s)
        	subscription_test.go:325: try 1 out of 10, message count was 1, not 0
        --- PASS: TestSB/TestSubscriptionClient/ReceiveOne (16.27s)
    --- PASS: TestSB/TestSubscriptionEntryUnmarshal (0.00s)
    --- PASS: TestSB/TestSubscriptionManagement (39.64s)
        --- PASS: TestSB/TestSubscriptionManagement/TestSubscriptionWithAutoDeleteOnIdle (4.85s)
        --- PASS: TestSB/TestSubscriptionManagement/TestSubscriptionWithRequiredSessions (5.95s)
        --- PASS: TestSB/TestSubscriptionManagement/TestSubscriptionWithDeadLetteringOnMessageExpiration (5.89s)
        --- PASS: TestSB/TestSubscriptionManagement/TestSubscriptionWithMessageTimeToLive (6.64s)
        --- PASS: TestSB/TestSubscriptionManagement/TestSubscriptionWithLockDuration (5.44s)
        --- PASS: TestSB/TestSubscriptionManagement/TestSubscriptionWithBatchedOperations (5.45s)
        --- PASS: TestSB/TestSubscriptionManagement/TestSubscriptionDefaultSettings (5.42s)
    --- PASS: TestSB/TestSubscriptionManagementWrites (8.32s)
        --- PASS: TestSB/TestSubscriptionManagementWrites/TestPutDefaultSubscription (0.52s)
    --- PASS: TestSB/TestSubscriptionUnmarshal (0.00s)
    --- PASS: TestSB/TestTopic (7.47s)
        --- PASS: TestSB/TestTopic/SimpleSend (7.47s)
    --- PASS: TestSB/TestTopicEntryUnmarshal (0.00s)
    --- PASS: TestSB/TestTopicManagement (21.34s)
        --- PASS: TestSB/TestTopicManagement/TopicWithExpress (1.80s)
        --- PASS: TestSB/TestTopicManagement/TopicWithDuplicateDetection (2.01s)
        --- PASS: TestSB/TestTopicManagement/TopicWithAutoDeleteOnIdle (2.25s)
        --- PASS: TestSB/TestTopicManagement/TopicWithTimeToLive (1.89s)
        --- PASS: TestSB/TestTopicManagement/TopicWithBatchOperations (2.19s)
        --- PASS: TestSB/TestTopicManagement/TopicWithMaxSizeInMegabytes (2.09s)
        --- PASS: TestSB/TestTopicManagement/DefaultTopicCreation (2.40s)
        --- PASS: TestSB/TestTopicManagement/TopicWithPartitioning (4.47s)
        --- PASS: TestSB/TestTopicManagement/TopicWithOrdering (2.24s)
    --- PASS: TestSB/TestTopicManagementReads (6.11s)
        --- PASS: TestSB/TestTopicManagementReads/TestGetTopic (0.30s)
        --- PASS: TestSB/TestTopicManagementReads/TestListTopics (0.80s)
    --- PASS: TestSB/TestTopicManagementWrites (1.90s)
        --- PASS: TestSB/TestTopicManagementWrites/TestPutDefaultTopic (1.90s)
    --- PASS: TestSB/TestTopicUnmarshal (0.00s)
```

marstr commented:
It may not look like it here, but I've merged this into `master` manually. :)
